deploy/single_user_deploy.sh

```diff
@@ -1,118 +1,150 @@
-        exit 1
-    fi
+    [[ "${visor_client_portal_rpc_in_visor}" != "${visor_client_portal_rpc_in_client}" ]] && echo -e "${ERR}mismatched VisorClientPortalService conf in ${CONF_FILE}, exiting ...${NC}" >&2 && exit 1
 
+    # for VisorKeeperRegistryService, Keeper->Visor
     visor_keeper_registry_rpc_in_visor=$(jq '.chrono_visor.VisorKeeperRegistryService.rpc' "${CONF_FILE}")
     visor_keeper_registry_rpc_in_keeper=$(jq '.chrono_keeper.VisorKeeperRegistryService.rpc' "${CONF_FILE}")
-    if [[ "${visor_keeper_registry_rpc_in_visor}" != "${visor_keeper_registry_rpc_in_keeper}" ]]
-    then
-        echo -e "${ERR}mismatched VisorKeeperRegistryService conf in ${CONF_FILE}, exiting ...${NC}"
-        exit 1
+    [[ "${visor_keeper_registry_rpc_in_visor}" != "${visor_keeper_registry_rpc_in_keeper}" ]] && echo -e "${ERR}mismatched VisorKeeperRegistryService conf in ${CONF_FILE}, exiting ...${NC}" >&2 && exit 1
+
+    # for VisorRegistryService, Grapher->Visor
+    visor_grapher_registry_rpc_in_grapher=$(jq '.chrono_grapher.VisorRegistryService.rpc' "${CONF_FILE}")
+    [[ "${visor_keeper_registry_rpc_in_visor}" != "${visor_grapher_registry_rpc_in_grapher}" ]] && echo -e "${ERR}mismatched VisorGrapherRegistryService conf in ${CONF_FILE}, exiting ...${NC}" >&2 && exit 1
+
+    # for KeeperGrapherDrainService, Keeper->Grapher
+    keeper_grapher_drain_rpc_in_keeper=$(jq '.chrono_keeper.KeeperGrapherDrainService.rpc' "${CONF_FILE}")
+    keeper_grapher_drain_rpc_in_grapher=$(jq '.chrono_grapher.KeeperGrapherDrainService.rpc' "${CONF_FILE}")
+    [[ "${keeper_grapher_drain_rpc_in_keeper}" != "${keeper_grapher_drain_rpc_in_grapher}" ]] && echo -e "${ERR}mismatched KeeperGrapherDrainService conf in ${CONF_FILE}, exiting ...${NC}" >&2 && exit 1
+
+    [[ "${verbose}" == "true" ]] && echo -e "${DEBUG}Check base conf file done${NC}"
+}
+
+check_op_validity() {
+    count=0
+    [[ $deploy == true ]] && ((count++))
+    [[ $stop == true ]] && ((count++))
+    [[ $kill == true ]] && ((count++))
+    [[ $reset == true ]] && ((count++))
+
+    if [[ $count -ne 1 ]]; then
+        echo -e "${ERR}Error: Please select exactly one operation in deploy (-d), stop (-s), kill (-k) and reset (-r).${NC}" >&2
+        usage
     fi
+}
 
-    if [[ "${verbose}" == "true" ]]
-    then
-        echo -e "${DEBUG}Check conf files done${NC}"
+check_recording_group_mapping() {
+    echo -e "${INFO}Checking recording group mapping ...${NC}"
+    local num_keepers
+    local num_recording_group
+
+    num_keepers=$(wc -l < ${KEEPER_HOSTS})
+    num_recording_group=${NUM_RECORDING_GROUP}
+    if [[ "${num_recording_group}" -lt 1 ]] || [[ "${num_recording_group}" -gt "${num_keepers}" ]] || (( num_keepers % num_recording_group != 0 )); then
+        echo -e "${ERR}NUM_RECORDING_GROUP must be greater than 0, less than or equal to the number of keepers (${num_keepers}), and a divisor of the number of keepers (${num_keepers}), exiting ...${NC}" >&2
```

Comment:
NUM_RECORDING_GROUP is dictated by the number of grapher processes . 
it's totally valid that (num_keepers % num_recording _group !=0)

Reply:
In this script, everything works in an automatic mode where all the conf files are generated from a single file like our `default_conf.json`. The `RecordingGroup` key in the JSON file is set during the generation. I totally agree that in more realistic case where users pass conf file for each process separately, num_recording_group may not be a divisor of num_keepers. But for current deployment, there is no good way to achieve that.